File: backend/src/main/java/com/autocare/serviceorder/controller/ServiceOrderController.java
package com.autocare.serviceorder.controller;

import com.autocare.serviceorder.dto.ServiceOrderRequestDTO;
import com.autocare.serviceorder.dto.ServiceOrderResponseDTO;
import com.autocare.serviceorder.dto.StatusUpdateRequestDTO;
import com.autocare.serviceorder.entity.ServiceOrderStatus;
import com.autocare.serviceorder.service.ServiceOrderService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * MECHANIC só é liberado em findById/updateStatus (para ver e diagnosticar a própria OS);
 * a restrição "somente a que foi atribuída a ele" é aplicada em
 * ServiceOrderService#assertCanAccessOrder, pois @PreAuthorize sozinho não sabe qual
 * mecânico está por trás do ID da URL.
 */
@RestController
@RequestMapping("/service-orders")
@RequiredArgsConstructor
@PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'RECEPTIONIST')")
public class ServiceOrderController {

    private final ServiceOrderService serviceOrderService;

    @PostMapping
    @PreAuthorize("hasAnyRole('ADMIN', 'RECEPTIONIST')")
    public ResponseEntity<ServiceOrderResponseDTO> create(@Valid @RequestBody ServiceOrderRequestDTO request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(serviceOrderService.create(request));
    }

    @GetMapping
    public ResponseEntity<List<ServiceOrderResponseDTO>> findAll() {
        return ResponseEntity.ok(serviceOrderService.findAll());
    }

    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'RECEPTIONIST', 'MECHANIC')")
    public ResponseEntity<ServiceOrderResponseDTO> findById(@PathVariable String id) {
        return ResponseEntity.ok(serviceOrderService.findByIdResponse(id));
    }

    @GetMapping("/client/{clientId}")
    public ResponseEntity<List<ServiceOrderResponseDTO>> findByClient(@PathVariable String clientId) {
        return ResponseEntity.ok(serviceOrderService.findByClient(clientId));
    }

    @GetMapping("/vehicle/{vehicleId}")
    public ResponseEntity<List<ServiceOrderResponseDTO>> findByVehicle(@PathVariable String vehicleId) {
        return ResponseEntity.ok(serviceOrderService.findByVehicle(vehicleId));
    }

    @GetMapping("/status/{status}")
    public ResponseEntity<List<ServiceOrderResponseDTO>> findByStatus(@PathVariable ServiceOrderStatus status) {
        return ResponseEntity.ok(serviceOrderService.findByStatus(status));
    }

    @PatchMapping("/{id}/status")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'MECHANIC')")
    public ResponseEntity<ServiceOrderResponseDTO> updateStatus(
            @PathVariable String id,
            @Valid @RequestBody StatusUpdateRequestDTO request) {
        return ResponseEntity.ok(serviceOrderService.updateStatus(id, request));
    }

    @PatchMapping("/{id}/mechanic")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'RECEPTIONIST')")
    public ResponseEntity<ServiceOrderResponseDTO> assignMechanic(
            @PathVariable String id,
            @RequestParam String mechanicId) {
        return ResponseEntity.ok(serviceOrderService.assignMechanic(id, mechanicId));
    }
}

File: backend/src/service-order/service-order.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  Patch,
  Post,
  Query,
  UseGuards,
} from "@nestjs/common";
import { ServiceOrderStatus } from "@prisma/client";
import { ServiceOrderService } from "./service-order.service";
import { ServiceOrderRequestDto, StatusUpdateRequestDto } from "./dto/service-order.dto";
import { Roles } from "../common/decorators/roles.decorator";
import { RolesGuard } from "../common/guards/roles.guard";
import { CurrentUser } from "../common/decorators/current-user.decorator";
import { Role } from "@prisma/client";
import { AuthUser } from "../auth/jwt.strategy";

// MECHANIC só é liberado em findById/updateStatus (para ver e diagnosticar a própria
// OS); a restrição "somente a que foi atribuída a ele" é aplicada em
// ServiceOrderService#assertCanAccessOrder, pois o RolesGuard sozinho não sabe qual
// mecânico está por trás do ID da URL.
@Controller("service-orders")
@UseGuards(RolesGuard)
@Roles(Role.ADMIN, Role.MANAGER, Role.RECEPTIONIST)
export class ServiceOrderController {
  constructor(private readonly serviceOrderService: ServiceOrderService) {}

  @Post()
  @Roles(Role.ADMIN, Role.RECEPTIONIST)
  @HttpCode(HttpStatus.CREATED)
  create(@Body() dto: ServiceOrderRequestDto) {
    return this.serviceOrderService.create(dto);
  }

  @Get()
  findAll() {
    return this.serviceOrderService.findAll();
  }

  @Get(":id")
  @Roles(Role.ADMIN, Role.MANAGER, Role.RECEPTIONIST, Role.MECHANIC)
  findById(@Param("id") id: string, @CurrentUser() user: AuthUser) {
    return this.serviceOrderService.findByIdResponse(id, user);
  }

  @Get("client/:clientId")
  findByClient(@Param("clientId") clientId: string) {
    return this.serviceOrderService.findByClient(clientId);
  }

  @Get("vehicle/:vehicleId")
  findByVehicle(@Param("vehicleId") vehicleId: string) {
    return this.serviceOrderService.findByVehicle(vehicleId);
  }

  @Get("status/:status")
  findByStatus(@Param("status") status: ServiceOrderStatus) {
    return this.serviceOrderService.findByStatus(status);
  }

  @Patch(":id/status")
  @Roles(Role.ADMIN, Role.MANAGER, Role.MECHANIC)
  updateStatus(
    @Param("id") id: string,
    @Body() dto: StatusUpdateRequestDto,
    @CurrentUser() user: AuthUser,
  ) {
    return this.serviceOrderService.updateStatus(id, dto, user);
  }

  @Patch(":id/mechanic")
  @Roles(Role.ADMIN, Role.MANAGER, Role.RECEPTIONIST)
  assignMechanic(@Param("id") id: string, @Query("mechanicId") mechanicId: string) {
    return this.serviceOrderService.assignMechanic(id, mechanicId);
  }
}
